src/model/types.rs

```diff
@@ -0,0 +1,82 @@
+use std::str::FromStr;
+
+/// Enumerates the two possible stone types
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Stone {
+    /// Stone placed by the current player.
+    ///
+    /// Defined as numerical value 1 in the piskwork protocol
+    Ally = 1,
+
+    /// Stone placed by the opponent player.
+    ///
+    /// Defined as numerical value 2 in the piskwork protocol
+    Opponent = 2,
+}
+
+/// Enumerates all possible states of a Gomoku cell
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CellContent {
+    /// No stone placed in this cell.
+    ///
+    /// Defined as numerical value 0 in the piskwork protocol
+    Empty = 0,
+
+    /// Cell contains a stone placed by the current player.
+    ///
+    /// See [`Stone::Ally`]
+    Ally = Stone::Ally as isize,
+
+    /// Cell contains a stone placed by the opponent player.
+    ///
+    /// See [`Stone::Opponent`]
+    Opponent = Stone::Opponent as isize,
+}
+
+/// Describes all the possible game types defined
+/// by the `INFO game_type` command
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum GameType {
+    /// AI plays against a human
+    Human,
+
+    /// AI plays against another AI
+    Brain,
+
+    /// AI plays in a tournament
+    Tournament,
+
+    /// AI plays in a network tournament
+    NetworkTournament,
+}
+
+impl FromStr for GameType {
+    type Err = &'static str;
```

Comment:
What does `&'static` do ?

Reply:
In Rust, the `'` character has a very important role: it specifies the lifetime of references. Contrary to C++, Rust tracks the lifetime of references at compile time and guarantees with its type system that no references to dropped objects can exist.

In this case, the `'static` lifetime indicates that the reference has static lifetime, meaning it is valid for the whole duration of the program, which is the case for string literals. So much so that `&'static str` can be considered to be the type of string literals